x509util/name.go

```diff
@@ -131,24 +149,44 @@ type DistinguishedName struct {
 	Value interface{}      `json:"value"`
 }
 
-func newDistinguisedNames(atvs []pkix.AttributeTypeAndValue) []DistinguishedName {
+// newExtraNames returns a list of DistinguishedName with the attributes not
+// present in attributeTypeNames.
+func newExtraNames(atvs []pkix.AttributeTypeAndValue) []DistinguishedName {
 	var extraNames []DistinguishedName
 	for _, atv := range atvs {
-		extraNames = append(extraNames, DistinguishedName{
-			Type:  ObjectIdentifier(atv.Type),
-			Value: atv.Value,
-		})
+		if _, ok := attributeTypeNames[atv.Type.String()]; !ok {
+			extraNames = append(extraNames, DistinguishedName{
+				Type:  ObjectIdentifier(atv.Type),
+				Value: atv.Value,
+			})
+		}
 	}
 	return extraNames
 }
 
+// fromDistinguisedNames converts a list of DistinguisedName to
+// []pkix.AttributeTypeAndValue. Note that this method has a special case to
+// encode the emailAddress deprecated field (1.2.840.113549.1.9.1).
 func fromDistinguisedNames(dns []DistinguishedName) []pkix.AttributeTypeAndValue {
```

Comment:
Needs update on the calling side too.

```suggestion
// fromDistinguishedNames converts a list of DistinguishedName to
// []pkix.AttributeTypeAndValue. Note that this method has a special case to
// encode the deprecated emailAddress field (1.2.840.113549.1.9.1).
func fromDistinguishedNames(dns []DistinguishedName) []pkix.AttributeTypeAndValue {
```

Reply:
Fixed with 487ffed